src/vm/contexts.rs

```diff
@@ -63,6 +83,246 @@ pub struct CallStack {
 
 pub type StackTrace = Vec<FunctionIdentifier>;
 
+impl AssetMap {
+    pub fn new() -> AssetMap {
+        AssetMap {
+            token_map: HashMap::new(),
+            asset_map: HashMap::new()
+        }
+    }
+
+    // This will get the next amount for a (principal, asset) entry in the asset table.
+    fn get_next_amount(&self, principal: &PrincipalData, asset: &AssetIdentifier, amount: i128) -> Result<i128> {
+        let current_amount = match self.token_map.get(principal) {
+            Some(principal_map) => *principal_map.get(&asset).unwrap_or(&0),
+            None => 0
+        };
+            
+        current_amount.checked_add(amount)
+            .ok_or(RuntimeErrorType::ArithmeticOverflow.into())
+    }
+
+    pub fn add_asset_transfer(&mut self, principal: &PrincipalData, asset: AssetIdentifier, transfered: Value) {
+        if !self.asset_map.contains_key(principal) {
+            self.asset_map.insert(principal.clone(), HashMap::new());
+        }
+
+        let principal_map = self.asset_map.get_mut(principal)
+            .unwrap(); // should always exist, because of checked insert above.
+
+        if principal_map.contains_key(&asset) {
+            principal_map.get_mut(&asset).unwrap().push(transfered); 
+        } else {
+            principal_map.insert(asset, vec![transfered]); 
+        }
+    }
+
+    pub fn add_token_transfer(&mut self, principal: &PrincipalData, asset: AssetIdentifier, amount: i128) -> Result<()> {
+        if amount < 0 {
+            panic!("Should never attempt to log a negative transfer.");
+        }
+
+        let next_amount = self.get_next_amount(principal, &asset, amount)?;
+
+        if !self.token_map.contains_key(principal) {
+            self.token_map.insert(principal.clone(), HashMap::new());
+        }
+
+        let principal_map = self.token_map.get_mut(principal)
+            .unwrap(); // should always exist, because of checked insert above.
+
+        principal_map.insert(asset, next_amount);
```

Comment:
If I did two `token-transfer!`s in a transaction on the same token type, won't this code cause my first `token-transfer!` to be overwritten by the second one?